### PR TITLE
Count from markup

### PR DESCRIPTION
**Pull request by k-bespalov**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -17,7 +17,7 @@ repositories {
 
 dependencies {
     compileOnly 'org.projectlombok:lombok:1.18.8'
-    compile 'org.telegram:telegrambots-spring-boot-starter:4.1.2'
+    compile 'org.telegram:telegrambots-spring-boot-starter:4.4.0.1'
```
Comment (yeplastun) [inline build.gradle]: why?

File: src/main/java/com/imbananko/tilly/MemeManager.java [hunk 2]
```diff
@@ -67,96 +72,143 @@ public void onUpdateReceived(Update update) {
...
-        "@" + update.getCallbackQuery().getMessage().getCaption().replaceFirst("Sender: ", "")
-        + ", поясни за мем";
-
-      Try.of(() -> execute(
-        new SendMessage()
-          .setChatId(message.getChatId())
-          .setReplyToMessageId(update.getCallbackQuery().getMessage().getMessageId())
-          .setText(replyText)
+    return CompletableFuture.runAsync(() -> {
+      final var voteWasExisted =
```
Comment (yeplastun) [inline src/main/java/com/imbananko/tilly/MemeManager.java]: didVoteExists

File: src/main/java/com/imbananko/tilly/MemeManager.java [hunk 3]
```diff
@@ -67,96 +72,143 @@ public void onUpdateReceived(Update update) {
...
+        (currentStats.upVoters.contains(voterUsername) && voteValue.equals(UP))
+          || (currentStats.explainVoters.contains(voterUsername) && voteValue.equals(EXPLAIN))
+          || (currentStats.downVoters.contains(voterUsername) && voteValue.equals(DOWN));
+
+      if (voteWasExisted) {
+        voteRepository.delete(voteEntity);
+      } else {
+        voteRepository.insertOrUpdate(voteEntity);
```
Comment (yeplastun) [inline src/main/java/com/imbananko/tilly/MemeManager.java]: do we need this if? insertOrUpdate should do the job in both cases

File: src/main/java/com/imbananko/tilly/MemeManager.java [hunk 4]
```diff
@@ -67,96 +72,143 @@ public void onUpdateReceived(Update update) {
...
 
-  private static InlineKeyboardMarkup createMarkup(HashMap<VoteEntity.Value, Long> stats) {
+  private static InlineKeyboardMarkup createMarkup(Statistics stats) {
     return new InlineKeyboardMarkup().setKeyboard(
-      List.of(
-        List.of(
-          createVoteInlineKeyboardButton(UP, stats.getOrElse(UP, 0L)),
-          createVoteInlineKeyboardButton(EXPLAIN, stats.getOrElse(EXPLAIN, 0L)),
-          createVoteInlineKeyboardButton(DOWN, stats.getOrElse(DOWN, 0L)))
+      java.util.List.of(
```
Comment (yeplastun) [inline src/main/java/com/imbananko/tilly/MemeManager.java]: imports?

File: src/main/java/com/imbananko/tilly/MemeManager.java [hunk 5]
```diff
@@ -67,96 +72,143 @@ public void onUpdateReceived(Update update) {
...
+          createVoteInlineKeyboardButton(DOWN, stats.downVoters)
+        )
       )
     );
   }
 
-  private static InlineKeyboardButton createVoteInlineKeyboardButton(VoteEntity.Value voteValue, long voteCount) {
+  private static InlineKeyboardButton createVoteInlineKeyboardButton(VoteEntity.Value voteValue, Set<String> voters) {
+    var voteCount = voters.length();
+    var votersString = voters.mkString("[", ",", "]");
```
Comment (yeplastun) [inline src/main/java/com/imbananko/tilly/MemeManager.java]: why not just coma separated? easier to parse

File: src/main/java/com/imbananko/tilly/MemeManager.java [hunk 6]
```diff
@@ -67,96 +72,143 @@ public void onUpdateReceived(Update update) {
...
+
+  private static Statistics getStatisticsFromMarkup(Message message) {
+    final var buttons = message.getReplyMarkup().getKeyboard().get(0);
+    final var votersList = List.ofAll(buttons).map(it -> extractVoters(it.getCallbackData()));
+
+    return new Statistics(votersList.get(0), votersList.get(1), votersList.get(2));
+  }
+
+  private static Set<String> extractVoters(String data) {
+    final var votersString = data.split(" ")[1].replaceAll("\\[|\\]", "");
```
Comment (yeplastun) [inline src/main/java/com/imbananko/tilly/MemeManager.java]: no need for replacing here so

File: build.gradle (hunk 1 above)
Comment (k-bespalov) [reply build.gradle]: new version has methods for getting markup data

File: src/main/java/com/imbananko/tilly/MemeManager.java (hunk 2 above)
Comment (k-bespalov) [reply src/main/java/com/imbananko/tilly/MemeManager.java]: fixed

**Pull request merged**